pkg/inject/secretless.go

```diff
@@ -1,55 +1,120 @@
-            envVarFromConfigMap("CONJUR_AUTHN_LOGIN", conjurAuthConfigMapName))
-    }
-
-    return &PatchConfig{
-        Containers: []corev1.Container{
-            {
-                Name:            "secretless",
-                Image:           "cyberark/secretless-broker:latest",
-                Args:            []string{"-f", "/etc/secretless/secretless.yml"},
-                ImagePullPolicy: "Always",
-                VolumeMounts: []corev1.VolumeMount{
-                    {
-                        Name:      "secretless-config",
-                        ReadOnly:  true,
-                        MountPath: "/etc/secretless",
-                    },
-                },
-                Env: envvars,
-            },
-        },
-        Volumes: []corev1.Volume{
-            {
-                Name: "secretless-config",
-                VolumeSource: corev1.VolumeSource{
-                    ConfigMap: &corev1.ConfigMapVolumeSource{
-                        LocalObjectReference: corev1.LocalObjectReference{
-                            Name: secretlessConfigMapName,
-                        },
-                    },
-                },
-            },
-        },
-    }
+func generateSecretlessSidecarConfig(secretlessConfig, conjurConnConfigMapName, conjurAuthConfigMapName, ServiceAccountTokenVolumeName string) *PatchConfig {
+	envvars := []corev1.EnvVar{
+		envVarFromFieldPath("MY_POD_NAME", "metadata.name"),
+		envVarFromFieldPath("MY_POD_NAMESPACE", "metadata.namespace"),
+		envVarFromFieldPath("MY_POD_IP", "status.podIP"),
+	}
+
+	if conjurConnConfigMapName != "" || conjurAuthConfigMapName != "" {
+		envvars = append(envvars,
+			envVarFromConfigMap("CONJUR_VERSION", conjurConnConfigMapName),
+			envVarFromConfigMap("CONJUR_APPLIANCE_URL", conjurConnConfigMapName),
+			envVarFromConfigMap("CONJUR_AUTHN_URL", conjurConnConfigMapName),
+			envVarFromConfigMap("CONJUR_ACCOUNT", conjurConnConfigMapName),
+			envVarFromConfigMap("CONJUR_SSL_CERTIFICATE", conjurConnConfigMapName),
+			envVarFromConfigMap("CONJUR_AUTHN_LOGIN", conjurAuthConfigMapName))
+	}
+
+	// Allow configmgr#configspec in the SecretlessConfig annotation
+	var configMgr string
+	var configSpec string
+	var secretlessConfigMapName string
+	secretlessConfigPath := "/etc/secretless/secretless.yml"
+	volumes := []corev1.Volume{}
+
+	// Always add Service Account Token Volume Mount (SATVM)
+	// It shouldn't be sidecar-injector's responsibility to add the SATVM, we only
+	// do it here because the serviceaccount plugin which adds the SATVM is
+	// executed before this plugin injects the sidecar. KEP-36 will solve this
+	// ordering problem by re-running plugins after mutation. Then if we add a
+	// container, the serviceaccount plugin will reprocess the manifest and add the
+	// appropriate mount
+	//
+	// ** Remove me when KEP-36 lands **
+	// also remove common.getServiceAccountTokenVolumeName
+	// and calls to that in server.go
+	volumeMounts := []corev1.VolumeMount{
```

Comment:
Is it always the case that the SATVM doesn't happen for the injected sidecar ? I'm wondering if there are cases where the SATVM works as expected and what the impact of doubly adding SATVM would be (since I haven't spotted a check for it). Are volume mounts idempotent ?

Reply:
I'm pretty sure with the current implementation there aren't edge cases where the mount will exist as plugin ordering is hard coded. We'll need to revise when the upstream issue is fixed though.

Reply:
I'm thinking 
1. v_now has the bug
2. v_later will not have the bug
3. if the code conditionally adds the service account volume mount then it will be compatible with both v_now and v_later

Reply:
At this point in the code, we're creating a container from scratch, so the mount can't exist. If a duplicate mount get's created, it will be by an external process after this. I'm currently testing deliberately injecting a duplicate volume mount to see if it breaks tests.

Reply:
Creating duplicate SATVMs does cause failure. However I'm really not sure what to do about it. As pointed out previously we're creating a container from scratch so the volume mount will not already exist for us to check for it. It may be added in a future step/plugin, and hopefully that process will be idempotent. 

On that note [KEP-36](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md#mutating-plugin-ordering) Does say that mutating webhooks that get re-invoked will have to be idempotent. So any plugin that sees the pod spec after injection must be idempotent so it shouldn't add the SATVM if it's already there.

Reply:
You're absolutely right. Point 3 in my comment above doesn't make sense since we have no idea if the volume mount will be added after us or not